src/common/Journald.cc

```diff
@@ -0,0 +1,301 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#include "Journald.h"
+
+#include <endian.h>
+#include <fcntl.h>
+#include <memory>
+#include <string>
+#include <sys/mman.h>
+#include <sys/socket.h>
+#include <sys/uio.h>
+#include <sys/un.h>
+#include <syslog.h>
+#include <unistd.h>
+#include <fmt/format.h>
+#include <fmt/ostream.h>
+
+#include "include/ceph_assert.h"
+#include "common/LogEntry.h"
+#include "log/Entry.h"
+#include "log/SubsystemMap.h"
+
+
+namespace ceph::logging {
+
+namespace {
+const struct sockaddr_un sockaddr = {
+  AF_UNIX,
+  "/run/systemd/journal/socket",
+};
+
+ssize_t sendmsg_fd(int transport_fd, int fd)
+{
+  constexpr size_t control_len = CMSG_LEN(sizeof(int));
+  char control[control_len];
+  struct msghdr mh = {
+    .msg_name = (struct sockaddr*)&sockaddr,
+    .msg_namelen = sizeof(sockaddr),
+    .msg_control = &control,
+    .msg_controllen = control_len,
+  };
+  ceph_assert(transport_fd >= 0);
+
+  struct cmsghdr *cmsg = CMSG_FIRSTHDR(&mh);
+  cmsg->cmsg_level = SOL_SOCKET;
+  cmsg->cmsg_type = SCM_RIGHTS;
+  cmsg->cmsg_len = CMSG_LEN(sizeof(int));
+  *reinterpret_cast<int *>(CMSG_DATA(cmsg)) = fd;
+
+  return sendmsg(transport_fd, &mh, MSG_NOSIGNAL);
+}
+
+char map_prio(short ceph_prio)
+{
+  if (ceph_prio < 0)
+    return LOG_ERR;
+  if (ceph_prio == 0)
+    return LOG_WARNING;
+  if (ceph_prio < 5)
+    return LOG_NOTICE;
+  if (ceph_prio < 10)
+    return LOG_INFO;
+  return LOG_DEBUG;
```

Comment:
I've been undecided on the wisdom of spewing Ceph debug logs to the systemd journal (with cephadm deployed containers or otherwise). Is this really recommended practice?

The mapping you've done is interesting but not necessarily wrong. I think if we have `debug_ms=5` you will get some amusing spam in the journal which pretends to be important.

Reply:
> I've been undecided on the wisdom of spewing Ceph debug logs to the systemd journal

Why not. journald itself can do filtering both on collecting time and viewing time based on its structured metadata. And we are already doing this with the default cephadm deployment.

> I think if we have debug_ms=5 you will get some amusing spam in the journal which pretends to be important.

prio=5 is mapped to LOG_INFO, which is the default level if we just connect the stderr to journald. I think it is OK. Anyway, if someone don’t like some logs, they can do the filtering themselves. This can be easier compared with logging to files.

By the way, the journald driver of conmon log everything at LOG_ERROR level. I think this is bad.

Reply:
@liewegas and @sebastian-philipp what do you think? i think this change is a nice improvement. but i am not sure regarding @batrick 's question:

> Is this really recommended practice?

Reply:
As we log to journald instead of a file, we need to be able to write anything (including debugs) to journald that we would write to the files. 

Btw we have a similar function in Python already:

https://github.com/ceph/ceph/blob/3ba59e93e0ef38a9fa8c0e5685b4466e3215c1ff/src/pybind/mgr/mgr_module.py#L663-L679

Reply:
> As we log to journald instead of a file, we need to be able to write anything (including debugs) to journald that we would write to the files.

This is not necessarily true. We could have a debug log for true "debug" information and a status/error/audit log go to systemd journal. Merging them is convenient for programmers but not necessarily a good thing for users.

Reply:
Sending log messages to different targets based on the level is something that none of the logging code currently supports.  I'm not sure if it makes sense or not, but I think even if we decide to go down that road it's an orthogonal change to this one